Services/DeviceModels.cs

```diff
@@ -138,12 +140,17 @@ public class DeviceModels : IDeviceModels
         /// </summary>
         public async Task<DeviceModel> UpsertAsync(DeviceModel deviceModel)
         {
+            if (this.CheckDeviceModelExistence(deviceModel.Id))
+            {
+                throw new ConflictingResourceException("Device model with id '" + deviceModel.Id + "'already existed!");
+            }
+
             try
             {
                 var result = await this.customDeviceModels.UpsertAsync(deviceModel);
                 deviceModel.ETag = result.ETag;
             }
-            catch(ConflictingResourceException exception)
+            catch (ConflictingResourceException exception)
             {
                 this.log.Error("Unable to update deivce model :'", () => new { exception });
```

Comment:
typo

Reply:
also, exception messages should not end with ":" as if the data would be concatenated. I'd change the code to:
```
this.log.Error("Unable to update device model'", () => new { exception });
```

Reply:
Done.